test/lib/y2firewall/clients/installation_finish_test.rb

```diff
@@ -47,48 +42,90 @@
+    let(:enable_firewall) { false }
+    let(:api) do
+      instance_double(Y2Firewall::Firewalld::Api, remove_service: true, add_service: true)
+    end
 
-        subject.write
-      end
+    before do
+      allow(proposal_settings).to receive("enable_firewall").and_return(enable_firewall)
+      allow(firewalld).to receive("api").and_return(api)
+      allow(firewalld).to receive("enable!")
+      allow(firewalld).to receive("disable!")
+      allow(proposal_settings).to receive("open_ssh").and_return(false)
+    end
 
-      it "removes the ssh service from the public zone if blocked in the proposal" do
-        expect(firewalld.api).to receive(:remove_service).with("public", "ssh")
+    it "enables the firewalld service if enabled in the proposal" do
+      allow(proposal_settings).to receive("enable_firewall").and_return(true)
+      expect(firewalld).to receive("enable!")
 
-        subject.write
+      subject.send(:configure_firewall)
+    end
+
+    it "disables the firewalld service if disabled in the proposal" do
+      expect(firewalld).to receive("disable!")
+
+      subject.send(:configure_firewall)
+    end
+
+    it "adds the ssh service to the default zone if opened in the proposal" do
+      expect(proposal_settings).to receive("open_ssh").and_return(true)
+      expect(api).to receive(:add_service).with(proposal_settings.default_zone, "ssh")
+
+      subject.send(:configure_firewall)
+    end
+
+    it "removes the ssh service from the default zone if blocked in the proposal" do
+      expect(api).to receive(:remove_service).with(proposal_settings.default_zone, "ssh")
+
+      subject.send(:configure_firewall)
+    end
+
+    context "when vnc is proposed to be open" do
+      let(:service_available) { true }
+
+      before do
+        allow(proposal_settings).to receive("open_vnc").and_return(true)
+        allow(api).to receive(:service_supported?).with("tigervnc").and_return(service_available)
       end
 
-      it "adds the vnc service to the public zone if opened in the proposal" do
-        allow(proposal_settings).to receive("open_vnc").and_return true
-        expect(firewalld.api).to receive(:add_service).with("public", "vnc-server")
+      context "and the tigervnc service definition is available" do
+        it "adds the tigervnc and the tigervnc-https services to the default zone" do
+          expect(api).to receive(:add_service).with(proposal_settings.default_zone, "tigervnc")
+          expect(api).to receive(:add_service)
+            .with(proposal_settings.default_zone, "tigervnc-https")
 
-        subject.write
+          subject.send(:configure_firewall)
+        end
       end
 
-      it "returns true" do
-        expect(subject.write).to eq true
+      context "and the tigervnc service definition is not available" do
+        let(:service_available) { false }
+        it "logs the error" do
+          expect(subject.log).to receive(:error).with(/service definition is not available/)
```

Comment:
logging is an additional tool for a programmer and as a such it should not be tested in my POV. Instead I'd expect test that the method returns ```false``` in this case

Reply:
It is an additional tool for a programmer but it is also part of the program and the workflow and it describes what the client does when the tigervnc service definition is not present.

It could raise an error, report an error or whatever. In this case it just logs de error.

Reply:
well I'm still not convinced. Logging is for programmer to get some meta information about the algorithm. However workflow should be driven by return values, exceptions, ... and these things should be covered.

It e.g. happens quite often to me that e.g. when debugging something I'm shifting, disabling and modifying log messages quite often. Each of these actions can influence the test. In the best case it will only break building the test package.

Reply:
So it seems there are two different things... whether it should be tested or not and whether we should fail just logging the error or reporting it.

Again will use an example from the same client that also logs the error instead of raise exceptions:

https://github.com/yast/yast-bootloader/blob/master/src/lib/bootloader/finish_client.rb#L91

Of course in our scenario the package should be present, and well I can omit the check of supported service just trying to enable the services and relying in the firewalld API for the error reporting or just add a report instead of the log if we think it deserve it.

About the test, as commented we are testing all the possible scenarios and we are ensuring we are logging it and it even takes more importance if as you say you modify the log messages or disable whatever. It is not the first time we do a typo logging and error or a warning and the customer founds the error first than us just because we didn't checked all the possible scenarios.

Reply:
I do not change logs texts that often, but when I do fixing the test in those cases is a no-brainer. Anyway, there is another important benefit IMHO: prevent introducing typos. Let's say that, in the future, I change this `log.error` for `log.warning`; or I introduce some interpolation in the string (and I do it wrong)... In those cases, I am covered and YaST won't crash.